toolchain/check/function.cpp

```diff
@@ -191,60 +191,88 @@ auto CheckFunctionTypeMatches(Context& context,
+                         SemIR::FunctionId prev_function_id,
+                         bool prev_is_import) -> bool {
   auto& prev_function = context.functions().Get(prev_function_id);
 
   if (!CheckRedecl(context, new_function, prev_function, {})) {
     return false;
   }
 
-  if (!is_definition) {
-    CARBON_DIAGNOSTIC(FunctionRedecl, Error,
-                      "Redundant redeclaration of function {0}.",
-                      SemIR::NameId);
-    CARBON_DIAGNOSTIC(FunctionPreviousDecl, Note, "Previously declared here.");
-    context.emitter()
-        .Build(node_id, FunctionRedecl, prev_function.name_id)
-        .Note(prev_function.decl_id, FunctionPreviousDecl)
-        .Emit();
-    // The diagnostic doesn't prevent a merge.
-    return true;
-  } else if (prev_function.definition_id.is_valid()) {
-    CARBON_DIAGNOSTIC(FunctionRedefinition, Error,
-                      "Redefinition of function {0}.", SemIR::NameId);
-    CARBON_DIAGNOSTIC(FunctionPreviousDefinition, Note,
-                      "Previously defined here.");
-    context.emitter()
-        .Build(node_id, FunctionRedefinition, prev_function.name_id)
-        .Note(prev_function.definition_id, FunctionPreviousDefinition)
-        .Emit();
-    // The second definition will be unused as a consequence of the error.
-    return true;
-  } else if (prev_function.is_extern) {
-    CARBON_DIAGNOSTIC(FunctionDefiningExtern, Error,
-                      "Cannot define `extern` function `{0}`.", SemIR::NameId);
-    CARBON_DIAGNOSTIC(FunctionPreviousExternDecl, Note,
-                      "Previously declared `extern` here.");
-    context.emitter()
-        .Build(node_id, FunctionDefiningExtern, prev_function.name_id)
-        .Note(prev_function.decl_id, FunctionPreviousExternDecl)
-        .Emit();
-    // The diagnostic doesn't prevent a merge.
-    return true;
-  }
+  CheckIsAllowedRedecl(context, node_id, new_function, new_is_definition,
+                       prev_function, prev_is_import);
 
-  // Track the signature from the definition, so that IDs in the body
-  // match IDs in the signature.
-  prev_function.definition_id = new_function.definition_id;
-  prev_function.implicit_param_refs_id = new_function.implicit_param_refs_id;
-  prev_function.param_refs_id = new_function.param_refs_id;
-  prev_function.return_type_id = new_function.return_type_id;
-  prev_function.return_slot_id = new_function.return_slot_id;
+  if (new_is_definition) {
+    // Track the signature from the definition, so that IDs in the body
+    // match IDs in the signature.
+    prev_function.definition_id = new_function.definition_id;
+    prev_function.implicit_param_refs_id = new_function.implicit_param_refs_id;
+    prev_function.param_refs_id = new_function.param_refs_id;
+    prev_function.return_type_id = new_function.return_type_id;
+    prev_function.return_slot_id = new_function.return_slot_id;
+  }
   return true;
```

Comment:
Should we also set `prev_function.is_extern` to `new_function.is_extern` here? Once we've seen a local declaration, I think it should no longer matter whether a previous (eg, imported) declaration was `extern`.